tests/many_to_many/tests.py

```diff
@@ -554,3 +556,28 @@ def test_inherited_models_selects(self):
             ]
         )
         self.assertQuerysetEqual(b.publications.all(), ['<Publication: Science Weekly>'])
+
+    @skipUnlessDBFeature('supports_foreign_keys')
+    def test_count_and_exists_query(self):
+        """
+        #29725 - Calling count() and exists() on a many to many relation
+        should not generate sql including unnecessary JOIN.
+        """
+        article = NullableTargetArticle.objects.create(headline='Python is good')
+        NullablePublicationThrough.objects.create(article=article, publication=None)
+        NullablePublicationThrough.objects.create(article=article, publication=self.p1)
+
+        with CaptureQueriesContext(connection) as captured_query:
+            self.assertEqual(self.a1.publications.count(), 1)
+            self.assertNotIn('JOIN', captured_query[0]['sql'])
+
+            self.assertEqual(article.publications.count(), 1)
+            self.assertNotIn('JOIN', captured_query[1]['sql'])
+            self.assertIn('"publication_id" IS NOT NULL', captured_query[1]['sql'])
```

Comment:
The fact only a single result is returned is a strong enough assertion here. Some database backend could translate `__isnull` to some different SQL.